tools/verifier-helper/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+use cw_utils::PaymentError;
+use lavs_apis::{
+    id::TaskId,
+    interfaces::voting::{TotalPowerResponse, VotingPowerResponse},
+    tasks::{TaskQueryMsg, TaskStatus, TaskStatusResponse},
+    verifier_simple::{OperatorVote, TaskMetadata},
+};
+
+use lavs_apis::interfaces::voting::QueryMsg as OperatorQueryMsg;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum VerifierError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("{0}")]
+    Payment(#[from] PaymentError),
+
+    #[error("Invalid percentage, must be between 1 and 100")]
+    InvalidPercentage,
+
+    #[error("Operator tried to vote twice: {0}")]
+    OperatorAlreadyVoted(String),
+
+    #[error("Task expired. Cannot vote on it")]
+    TaskExpired,
+
+    #[error("Task already completed. Cannot vote on it")]
+    TaskAlreadyCompleted,
+
+    #[error("Unauthorized")]
+    Unauthorized,
+    // Add any other custom errors you like here.
+    // Look at https://docs.rs/thiserror/1.0.21/thiserror/ for details.
+}
+
+/// Does all checks to ensure the voter is valid and has not voted yet.
+/// Also checks the task is valid and still open.
+/// Returns the metadata for the task (creating it if first voter), along with the voting power of this operator.
+///
+/// We do not want to error if an operator votes for a task that is already completed (due to race conditions).
+/// In that case, just return None and exit early rather than error.
+#[allow(clippy::too_many_arguments)]
+pub fn ensure_valid_vote(
+    mut deps: DepsMut,
+    env: &Env,
+    task_queue: &Addr,
+    task_id: TaskId,
+    operator: &Addr,
+    vote: Option<OperatorVote>,
+    metadata: &Option<TaskMetadata>,
+    fraction_percent: u32,
+    operators_addr: &Addr,
+) -> Result<Option<(TaskMetadata, Uint128)>, VerifierError> {
+    // Operator has not submitted a vote yet
+    if vote.is_some() {
+        return Err(VerifierError::OperatorAlreadyVoted(operator.to_string()));
+    }
+
+    // get config for future queries
+
+    // Load task info, or create it if not there
+    // Error here means the contract is in expired or completed, return None rather than error
+    let metadata = match load_or_initialize_metadata(
+        deps.branch(),
+        env,
+        operators_addr,
+        task_queue,
+        task_id,
+        metadata,
+        fraction_percent,
+    ) {
+        Ok(x) => x,
+        Err(_) => return Ok(None),
+    };
+
+    // Get the operators voting power at time of vote creation
+    let power: VotingPowerResponse = deps.querier.query_wasm_smart(
+        operators_addr.to_string(),
+        &OperatorQueryMsg::VotingPowerAtHeight {
+            address: operator.to_string(),
+            height: Some(metadata.created_height),
+        },
+    )?;
+    if power.power.is_zero() {
+        return Err(VerifierError::Unauthorized);
+    }
+
+    Ok(Some((metadata, power.power)))
+}
+
+fn load_or_initialize_metadata(
```

Comment:
Note, this doesn't actually load it. You could rename it, but actually, I would love to handle the `TASK` Map inside this package, like we do in [cw_controllers](https://docs.rs/cw-controllers/latest/cw_controllers/struct.Claims.html).

You can make the map and wrap it in a struct with these higher level methods. And then read/write it here

Reply:
wow, this one made me sweat at the start, hopefully is done in b9c3011c0157